test/hpke_test.c

```diff
@@ -0,0 +1,1432 @@
+        ksinfo, sizeof(ksinfo),
+        NULL, 0, /* no auth */
+        NULL, 0, NULL /* PSK stuff */
+    };
+    const TEST_EXPORTDATA exportdata[] = {
+        { NULL, 0, fourth_export1, sizeof(fourth_export1) },
+        { fourth_context2, sizeof(fourth_context2),
+          fourth_export2, sizeof(fourth_export2) },
+        { fourth_context3, sizeof(fourth_context3),
+          fourth_export3, sizeof(fourth_export3) },
+    };
+    return do_testhpke(&basedata, NULL, 0,
+                       exportdata, OSSL_NELEM(exportdata));
+}
+
+/*
+ * Randomly toss a coin
+ */
+static unsigned char rb = 0;
+#define COIN_IS_HEADS (RAND_bytes_ex(testctx, &rb, 1, 10) && rb % 2)
+
+/* tables of HPKE modes and suite values */
+static int hpke_mode_list[] = {
+    OSSL_HPKE_MODE_BASE,
+    OSSL_HPKE_MODE_PSK,
+    OSSL_HPKE_MODE_AUTH,
+    OSSL_HPKE_MODE_PSKAUTH
+};
+static uint16_t hpke_kem_list[] = {
+    OSSL_HPKE_KEM_ID_P256,
+    OSSL_HPKE_KEM_ID_P384,
+    OSSL_HPKE_KEM_ID_P521,
+    OSSL_HPKE_KEM_ID_X25519,
+    OSSL_HPKE_KEM_ID_X448
+};
+static uint16_t hpke_kdf_list[] = {
+    OSSL_HPKE_KDF_ID_HKDF_SHA256,
+    OSSL_HPKE_KDF_ID_HKDF_SHA384,
+    OSSL_HPKE_KDF_ID_HKDF_SHA512
+};
+static uint16_t hpke_aead_list[] = {
+    OSSL_HPKE_AEAD_ID_AES_GCM_128,
+    OSSL_HPKE_AEAD_ID_AES_GCM_256,
+    OSSL_HPKE_AEAD_ID_CHACHA_POLY1305
+};
+
+/* strings that can be used, with names or IANA codepoints */
+static char *kem_str_list[] = {
+    "P-256", "P-384", "P-521", "x25519", "x448",
+    "0x10", "0x11", "0x12", "0x20", "0x21",
+    "16", "17", "18", "32", "33"
+};
+static char *kdf_str_list[] = {
+    "hkdf-sha256", "hkdf-sha384", "hkdf-sha512",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3"
+};
+static char *aead_str_list[] = {
+    "aes-128-gcm", "aes-256-gcm", "chacha20-poly1305",
+    "0x1", "0x01", "0x2", "0x02", "0x3", "0x03",
+    "1", "2", "3"
+};
+/* table of bogus strings that better not work */
+static char *bogus_suite_strs[] = {
+    "3,33,3",
+    "bogus,bogus,bogus",
+    "bogus,33,3,1,bogus",
+    "bogus,33,3,1",
+    "bogus,bogus",
+    "bogus",
+    /* and one in the wrong order */
+    "aes-256-gcm,hkdf-sha512,x25519"
+};
+
+/**
+ * @brief round-trips, generating keys, encrypt and decrypt
+ *
+ * This iterates over all mode and ciphersuite options trying
+ * a key gen, encrypt and decrypt for each. The aad, info, and
+ * seq inputs are randomly set or omitted each time. EVP and
+ * non-EVP key generation are randomly selected.
+ *
+ * @return 1 for success, other otherwise
+ */
+static int test_hpke_modes_suites(void)
+{
+    int overallresult = 1;
+    int mind = 0; /* index into hpke_mode_list */
+    int kemind = 0; /* index into hpke_kem_list */
+    int kdfind = 0; /* index into hpke_kdf_list */
+    int aeadind = 0; /* index into hpke_aead_list */
+
+    /* iterate over the different modes */
+    for (mind = 0; mind != (sizeof(hpke_mode_list) / sizeof(int)); mind++) {
```

Comment:
This seems better implemented as a proper test iteration, i.e. add an int arg to `test_hpke_modes_suites` and use `ADD_ALL_TESTS` instead of `ADD_TEST` in setup. Doing it this way means you can rerun the test and specify a specific test and iteration number via the "-test" and "-iter" command line flags to the test executable (useful for debugging).

Separately, we should use `OSSL_NELEM` to determine the number of elements in `hpke_mode_list`

Reply:
``OSSL_NELEM`` bit done, will come back to the other

Reply:
Had a look at that and I reckon it's not that worthwhile in the end. That function has nested loops for mode/kem/kdf and aead and there's between 3 and 5 options for each (today). So if we added that int arg we'd either have to pick which of (mode, kem, kdf, aead) it represents or we'd need to flatten the combination of mode/kem/kdf/aead down to a single int. I'm not at all sure either offers much utility in the end, compared to running ``test/hpke_test`` in gdb and seeing where it fails, which is arguably easier than knowing to run with ``-test`` and ``-iter`` (at least for me - the fact ``-iter`` is possible is news to me just now:-)

So, I'd argue to leave as-is I think.

Reply:
> So, I'd argue to leave as-is I think. 

ok.